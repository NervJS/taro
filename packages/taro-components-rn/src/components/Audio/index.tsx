/**
 * 音频组件
 *
 * @see https://developers.weixin.qq.com/miniprogram/dev/component/audio.html#audio
 *
 * - id
 * ✔ src
 * ✔ loop
 * ✘ controls
 * ✔ poster
 * ✔ name
 * ✔ author
 * ✔ onError
 * ✔ onPlay
 * ✔ onPause
 * ✔ onTimeUpdate
 * ✔ onEnded
 */

import * as React from 'react'
import {
  TouchableWithoutFeedback,
  ActivityIndicator,
  Text,
  View,
  ImageBackground,
  NetInfo
} from 'react-native'
import { Audio } from 'expo'
import { MaterialIcons } from '@expo/vector-icons'
import utils from '../../utils'
import styles from './styles'

const soundObject = new Audio.Sound()

const enum PLAYBACK_STATES {
  LOADING = 'LOADING',
  PLAYING = 'PLAYING',
  PAUSED = 'PAUSED',
  BUFFERING = 'BUFFERING',
  ERROR = 'ERROR',
  ENDED = 'ENDED'
}

const enum MediaErrorCode {
  MEDIA_ERR_ABORTED = 1,
  MEDIA_ERR_NETWORK,
  MEDIA_ERR_DECODE,
  MEDIA_ERR_SRC_NOT_SUPPORTED
}

export interface Props {
  src: string;
  loop?: boolean;
  controls?: boolean;
  poster?: string;
  name?: string;
  author?: string;
  onError?(event: { errMsg: MediaErrorCode }): void;
  onPlay?(): void;
  onPause?(): void;
  onTimeUpdate?(event: { currentTime: number, duration: number }): void;
  onEnded?(): void;
}

export interface State {
  networkState: string;
  playbackState: PLAYBACK_STATES;
  lastPlaybackStateUpdate: number;
  playbackInstancePosition: number;
  playbackInstanceDuration: number;
  shouldPlay: boolean;
}

class _Audio extends React.Component<Props, State> {
  static defaultProps = {
    name: '未知音频',
    author: '未知作者',
    onError: utils.noop,
    onPlay: utils.noop,
    onPause: utils.noop,
    onTimeUpdate: utils.noop,
    onEnded: utils.noop,
  }
  state: State = {
    networkState: '',
    playbackState: PLAYBACK_STATES.LOADING,
    lastPlaybackStateUpdate: Date.now(),
    playbackInstancePosition: 0,
    playbackInstanceDuration: 0,
    shouldPlay: false
  }

  getSource() {
    const { src } = this.props
    return typeof src === 'string' ? { uri: src } : src
  }

  _onConnectionChange(connectionInfo: any) {
    this.setState({ networkState: connectionInfo.type })
  }

  _setupNetInfoListener() {
    NetInfo.getConnectionInfo().then((connectionInfo: any) => {
      this.setState({ networkState: connectionInfo.type })
    })
    NetInfo.addEventListener(
      'connectionChange',
      this._onConnectionChange.bind(this)
    )
  }

  _setPlaybackState = (playbackState: PLAYBACK_STATES) => {
    if (playbackState === PLAYBACK_STATES.PLAYING) {
      this.props.onTimeUpdate({ currentTime: this.state.playbackInstancePosition, duration: this.state.playbackInstanceDuration })
    }
    if (this.state.playbackState !== playbackState) {
      this.setState({ playbackState, lastPlaybackStateUpdate: Date.now() })
      if (playbackState === PLAYBACK_STATES.PLAYING) {
        this.props.onPlay()
      } else if (playbackState === PLAYBACK_STATES.PAUSED) {
        this.props.onPause()
      } else if (playbackState === PLAYBACK_STATES.ENDED) {
        this.props.onEnded()
      }
    }
  }

  _onPlaybackStatusUpdate = (playbackStatus: any) => {
    if (!playbackStatus.isLoaded) {
      if (playbackStatus.error) {
        this._setPlaybackState(PLAYBACK_STATES.ERROR)
        this.props.onError({ errMsg: MediaErrorCode.MEDIA_ERR_NETWORK })
      }
      return
    }
    this.setState({
      playbackInstancePosition: playbackStatus.positionMillis,
      playbackInstanceDuration: playbackStatus.durationMillis,
      shouldPlay: playbackStatus.shouldPlay,
    })
    if (this.state.playbackState !== PLAYBACK_STATES.ENDED) {
      if (playbackStatus.didJustFinish && !playbackStatus.isLooping) {
        this._setPlaybackState(PLAYBACK_STATES.ENDED)
      } else {
        if (this.state.networkState === 'none' && playbackStatus.isBuffering) {
          this._setPlaybackState(PLAYBACK_STATES.ERROR)
        } else {
          this._setPlaybackState(playbackStatus.isPlaying ? PLAYBACK_STATES.PLAYING : (playbackStatus.isBuffering ? PLAYBACK_STATES.BUFFERING : PLAYBACK_STATES.PAUSED))
        }
      }
    }
  }

  _getMMSSFromMillis = (millis: number) => {
    const totalSeconds = millis / 1000
    const seconds = Math.floor(totalSeconds % 60)
    const minutes = Math.floor(totalSeconds / 60)

    const padWithZero = (number: number) => {
      const string = number.toString()
      if (number < 10) {
        return '0' + string
      }
      return string
    }
    return padWithZero(minutes) + ':' + padWithZero(seconds)
  }

  _getPercent = () => {
    if (
      soundObject != null &&
      this.state.playbackInstancePosition != null &&
      this.state.playbackInstanceDuration != null
    ) {
      return (
        this.state.playbackInstancePosition /
        this.state.playbackInstanceDuration
      ) * 100
    }
    return 0
  }

  onToggle = async () => {
    switch (this.state.playbackState) {
      case PLAYBACK_STATES.ENDED:
        soundObject.setStatusAsync({ shouldPlay: true, positionMillis: 0 })
        break
      case PLAYBACK_STATES.PLAYING:
        soundObject.setStatusAsync({ shouldPlay: false })
        break
      case PLAYBACK_STATES.PAUSED:
        soundObject.setStatusAsync({ shouldPlay: true })
        break
    }
  }

  async componentDidMount() {
    this._setupNetInfoListener()
    soundObject.setOnPlaybackStatusUpdate(this._onPlaybackStatusUpdate)
    try {
      await soundObject.loadAsync(this.getSource(), {
        isLooping: !!this.props.loop,
      })
    } catch (err) {
      // ERR
    }
  }

  componentWillUnmount() {
    NetInfo.removeEventListener(
      'connectionChange',
      this._onConnectionChange.bind(this)
    )
    soundObject.unloadAsync()
  }

  render() {
    const { playbackState } = this.state
    return (
      <View style={styles.wrapper}>
        <View style={styles.container}>
          <TouchableWithoutFeedback onPress={this.onToggle}>
            <View style={styles.poster}>
              <ImageBackground source={{ uri: this.props.poster }} style={styles.posterImg}>
                {(playbackState === PLAYBACK_STATES.PAUSED || playbackState === PLAYBACK_STATES.ENDED) && <MaterialIcons
                  name={'play-circle-outline'}
                  size={24}
                  color="white"
                  style={styles.playIcon}
                />}
                {playbackState === PLAYBACK_STATES.PLAYING && <MaterialIcons
                  name={'pause-circle-outline'}
                  size={24}
                  color="white"
                  style={styles.playIcon}
                />}
                {(playbackState === PLAYBACK_STATES.LOADING ||
                  playbackState === PLAYBACK_STATES.BUFFERING ||
                  playbackState === PLAYBACK_STATES.ERROR) &&
                  <ActivityIndicator color="white" size={'small'} />}
              </ImageBackground>
            </View>
          </TouchableWithoutFeedback>
          <View style={styles.info}>
            <View style={styles.detail}>
              <View style={styles.name}><Text style={styles.nameText}>{this.props.name}</Text></View>
              <View style={styles.author}><Text style={styles.authorText}>{this.props.author}</Text></View>
            </View>
            <View style={styles.time}><Text style={styles.timeText}>{this._getMMSSFromMillis(this.state.playbackInstancePosition)}</Text></View>
          </View>
        </View>
        <View style={styles.progressBar}>
          <View style={[styles.progressBarTint, {
            width: `${this._getPercent()}%`
          }]}>
          </View>
        </View>
      </View>
    )
  }
}

export default _Audio
