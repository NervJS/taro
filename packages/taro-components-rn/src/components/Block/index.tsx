import { _View as _Block } from '../View/index'

// _Block.displayName = '_Block'

export { _Block }
export default _Block
