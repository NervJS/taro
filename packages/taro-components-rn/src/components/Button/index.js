/**
 * ✔ size
 * ✔ type
 * ✔ plain
 * ✔ disabled
 * ✔ loading
 * ✔ formType(form-type)
 * ✔ hoverStyle: Convert hoverClass to hoverStyle.
 * ✘ hoverStopPropagation
 * ✔ hoverStartTime
 * ✔ hoverStayTime
 * ✔ onClick
 * - open-type
 * - lang
 * - bindgetuserinfo
 * - session-from
 * - send-message-title
 * - send-message-path
 * - send-message-img
 * - show-message-card
 * - bindcontact
 * - bindgetphonenumber
 * - app-parameter
 * - binderror
 * - bindopensetting
 *
 * Loading icon (svg): data:image/svg+xml;charset=utf8, <svg xmlns='http://www.w3.org/2000/svg' width='120' height='120' viewBox='0 0 100 100'><path fill='none' d='M0 0h100v100H0z'/><rect width='7' height='20' x='46.5' y='40' fill='%23E9E9E9' rx='5' ry='5' transform='translate(0 -30)'/><rect width='7' height='20' x='46.5' y='40' fill='%23989697' rx='5' ry='5' transform='rotate(30 105.98 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%239B999A' rx='5' ry='5' transform='rotate(60 75.98 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23A3A1A2' rx='5' ry='5' transform='rotate(90 65 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23ABA9AA' rx='5' ry='5' transform='rotate(120 58.66 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23B2B2B2' rx='5' ry='5' transform='rotate(150 54.02 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23BAB8B9' rx='5' ry='5' transform='rotate(180 50 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23C2C0C1' rx='5' ry='5' transform='rotate(-150 45.98 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23CBCBCB' rx='5' ry='5' transform='rotate(-120 41.34 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23D2D2D2' rx='5' ry='5' transform='rotate(-90 35 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23DADADA' rx='5' ry='5' transform='rotate(-60 24.02 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23E2E2E2' rx='5' ry='5' transform='rotate(-30 -5.98 65)'/></svg>
 * Loading icon for warn type (svg): data:image/svg+xml;charset=utf8, %3Csvg xmlns='http://www.w3.org/2000/svg' width='120' height='120' viewBox='0 0 100 100'%3E%3Cpath fill='none' d='M0 0h100v100H0z'/%3E%3Crect xmlns='http://www.w3.org/2000/svg' width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.56)' rx='5' ry='5' transform='translate(0 -30)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.5)' rx='5' ry='5' transform='rotate(30 105.98 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.43)' rx='5' ry='5' transform='rotate(60 75.98 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.38)' rx='5' ry='5' transform='rotate(90 65 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.32)' rx='5' ry='5' transform='rotate(120 58.66 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.28)' rx='5' ry='5' transform='rotate(150 54.02 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.25)' rx='5' ry='5' transform='rotate(180 50 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.2)' rx='5' ry='5' transform='rotate(-150 45.98 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.17)' rx='5' ry='5' transform='rotate(-120 41.34 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.14)' rx='5' ry='5' transform='rotate(-90 35 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.1)' rx='5' ry='5' transform='rotate(-60 24.02 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.03)' rx='5' ry='5' transform='rotate(-30 -5.98 65)'/%3E%3C/svg%3E
 *
 * @flow
 */

import * as React from 'react'
import {
  TouchableOpacity,
  Text,
  View,
  Image,
  Animated,
  Easing,
  StyleSheet,
} from 'react-native'
import styles from './styles'

type Props = {
  style?: StyleSheet.Styles,
  children?: React.Node,
  size: 'default' | 'mini',
  type: 'primary' | 'default' | 'warn',
  plain?: boolean,
  disabled?: boolean,
  loading?: boolean,
  formType?: 'submit' | 'reset',
  hoverStyle?: StyleSheet.Styles,
  // hoverStopPropagation: boolean,
  hoverStartTime?: number,
  hoverStayTime?: number,
  onClick?: Function,
}

type State = {
  valve: Animated.Value,
  isHover: boolean,
}

class _Button extends React.Component<Props, State> {
  props: Props
  state: State = {
    valve: new Animated.Value(0),
    isHover: false
  }
  $touchable: React.ElementRef<TouchableOpacity>

  static defaultProps = {
    size: 'default',
    type: 'default',
    hoverStartTime: 20,
    hoverStayTime: 70
  }

  _simulateNativePress = () => {
    this.$touchable.touchableHandlePress()
  }

  animate = () => {
    if (!this.props.loading) return

    Animated.sequence([
      Animated.timing(this.state.valve, {
        toValue: 1,
        easing: Easing.linear,
        duration: 1000
      }),
      Animated.timing(this.state.valve, {
        toValue: 0,
        duration: 0
      })
    ]).start(() => { this.animate() })
  }

  onPress = () => {
    const { disabled, onClick } = this.props
    !disabled && onClick && onClick()
  }

  onPressIn = () => {
    this.setState({ isHover: true })
  }

  onPressOut = () => {
    this.setState({ isHover: false })
  }

  componentDidMount () {
    this.animate()
  }

  componentDidUpdate (prevProps: Props) {
    if (!prevProps.loading && this.props.loading) {
      this.animate()
    }
  }

  // eslint-disable-next-line camelcase
  // UNSAFE_componentWillReceiveProps (nextProps: Props) {
  //   if (!this.props.loading && nextProps.loading) {
  //     this.animate()
  //   }
  // }

  render () {
    const {
      style,
      children,
      size,
      type,
      plain,
      disabled,
      loading,
      hoverStyle,
      hoverStartTime,
      hoverStayTime,
    } = this.props

    const isDefaultSize = size === 'default'
    const isDefaultType = type === 'default'
    const themeColorMap = {
      default: ['#F8F8F8', '#f7f7f7'],
      primary: ['#1AAD19', '#9ED99D'],
      warn: ['#E64340', '#EC8B89'],
    }
    // Use themeColorMap normally as PLAIN is false (by default),
    // otherwise use rgb(53,53,53) for plain-default-type particularly.
    const themeColor = plain && isDefaultType ? `rgba(53,53,53,${disabled ? 0.6 : 1})` : themeColorMap[type][disabled ? 1 : 0]
    const backgroundColor = plain ? 'transparent' : themeColor
    const borderStyle = plain && { borderWidth: 1, borderColor: themeColor }
    const textColor = plain
      ? themeColor
      : (isDefaultType ? `rgba(0,0,0,${disabled ? 0.3 : 1})` : `rgba(255,255,255,${disabled ? 0.6 : 1})`)

    const rotateDeg = this.state.valve.interpolate({
      inputRange: [0, 1],
      outputRange: ['0deg', '360deg']
    })

    return (
      <TouchableOpacity
        // activeOpacity={disabled ? 1 : 0.6}
        activeOpacity={1}
        delayPressIn={hoverStartTime}
        delayPressOut={hoverStayTime}
        onPress={this.onPress}
        onPressIn={this.onPressIn}
        onPressOut={this.onPressOut}
        ref={(touchable) => { this.$touchable = touchable }}
      >
        <View
          style={[
            styles.btn,
            !isDefaultSize && styles.btnMini,
            { backgroundColor },
            borderStyle,
            style,
            this.state.isHover && hoverStyle
          ]}
        >
          {loading && (
            <Animated.View
              style={[styles.loading, { transform: [{ rotate: rotateDeg }] }]}
            >
              <Image
                source={type === 'warn' ? require('../../assets/loading-warn.png') : require('../../assets/loading.png')}
                style={styles.loadingImg}
              />
            </Animated.View>)
          }
          {typeof children === 'string' ? <Text
            style={[
              styles.btnText,
              !isDefaultSize && styles.btnTextMini,
              { color: textColor }
            ]}
          >
            {children}
          </Text> : children}
        </View>
      </TouchableOpacity>
    )
  }
}

export default _Button
