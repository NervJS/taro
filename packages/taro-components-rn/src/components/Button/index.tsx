/**
 * ✔ size
 * ✔ type
 * ✔ plain
 * ✔ disabled
 * ✔ loading
 * ✔ formType (form-type)
 * ✔ hoverStyle: Convert hoverClass to hoverStyle.
 * ✘ hoverStopPropagation
 * ✔ hoverStartTime
 * ✔ hoverStayTime
 * ✔ onClick
 * - open-type
 * - lang
 * - bindgetuserinfo
 * - session-from
 * - send-message-title
 * - send-message-path
 * - send-message-img
 * - show-message-card
 * - bindcontact
 * - bindgetphonenumber
 * - app-parameter
 * - binderror
 * - bindopensetting
 *
 * Loading icon (svg): data:image/svg+xml;charset=utf8, <svg xmlns='http://www.w3.org/2000/svg' width='120' height='120' viewBox='0 0 100 100'><path fill='none' d='M0 0h100v100H0z'/><rect width='7' height='20' x='46.5' y='40' fill='%23E9E9E9' rx='5' ry='5' transform='translate(0 -30)'/><rect width='7' height='20' x='46.5' y='40' fill='%23989697' rx='5' ry='5' transform='rotate(30 105.98 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%239B999A' rx='5' ry='5' transform='rotate(60 75.98 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23A3A1A2' rx='5' ry='5' transform='rotate(90 65 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23ABA9AA' rx='5' ry='5' transform='rotate(120 58.66 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23B2B2B2' rx='5' ry='5' transform='rotate(150 54.02 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23BAB8B9' rx='5' ry='5' transform='rotate(180 50 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23C2C0C1' rx='5' ry='5' transform='rotate(-150 45.98 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23CBCBCB' rx='5' ry='5' transform='rotate(-120 41.34 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23D2D2D2' rx='5' ry='5' transform='rotate(-90 35 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23DADADA' rx='5' ry='5' transform='rotate(-60 24.02 65)'/><rect width='7' height='20' x='46.5' y='40' fill='%23E2E2E2' rx='5' ry='5' transform='rotate(-30 -5.98 65)'/></svg>
 * Loading icon for warn type (svg): data:image/svg+xml;charset=utf8, %3Csvg xmlns='http://www.w3.org/2000/svg' width='120' height='120' viewBox='0 0 100 100'%3E%3Cpath fill='none' d='M0 0h100v100H0z'/%3E%3Crect xmlns='http://www.w3.org/2000/svg' width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.56)' rx='5' ry='5' transform='translate(0 -30)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.5)' rx='5' ry='5' transform='rotate(30 105.98 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.43)' rx='5' ry='5' transform='rotate(60 75.98 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.38)' rx='5' ry='5' transform='rotate(90 65 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.32)' rx='5' ry='5' transform='rotate(120 58.66 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.28)' rx='5' ry='5' transform='rotate(150 54.02 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.25)' rx='5' ry='5' transform='rotate(180 50 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.2)' rx='5' ry='5' transform='rotate(-150 45.98 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.17)' rx='5' ry='5' transform='rotate(-120 41.34 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.14)' rx='5' ry='5' transform='rotate(-90 35 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.1)' rx='5' ry='5' transform='rotate(-60 24.02 65)'/%3E%3Crect width='7' height='20' x='46.5' y='40' fill='rgba(255,255,255,.03)' rx='5' ry='5' transform='rotate(-30 -5.98 65)'/%3E%3C/svg%3E
 */

import * as React from 'react'
import {
  TouchableWithoutFeedback,
  Text,
  View,
  Image,
  Animated,
  Easing,
  StyleProp,
  ViewStyle,
  GestureResponderEvent
} from 'react-native'
import styles from './styles'
import { noop } from '../../utils'
import { ButtonProps, ButtonState } from './PropsType'

class _Button extends React.Component<ButtonProps, ButtonState> {
  static defaultProps = {
    size: 'default',
    type: 'default',
    hoverStartTime: 20,
    hoverStayTime: 70
  }

  $touchable = React.createRef<TouchableWithoutFeedback>()

  state: ButtonState = {
    valve: new Animated.Value(0),
    isHover: false
  }

  animate = (): void => {
    if (!this.props.loading) return

    Animated.sequence([
      Animated.timing(this.state.valve, {
        toValue: 1,
        easing: Easing.linear,
        duration: 1000
      }),
      Animated.timing(this.state.valve, {
        toValue: 0,
        duration: 0
      })
    ]).start(() => { this.animate() })
  }

  onPress = (): void => {
    const { disabled, onClick = noop } = this.props
    !disabled && onClick()
  }

  onPressIn = (): void => {
    this.setState({ isHover: true })
  }

  onPressOut = (): void => {
    this.setState({ isHover: false })
  }

  _simulateNativePress = (evt: GestureResponderEvent): void => {
    const node = this.$touchable.current
    node && node.touchableHandlePress(evt)
  }

  componentDidMount () {
    this.animate()
  }

  componentDidUpdate (prevProps: ButtonProps) {
    if (!prevProps.loading && this.props.loading) {
      this.animate()
    }
  }

  render () {
    const {
      style,
      children,
      size,
      type,
      plain,
      disabled,
      loading,
      hoverStyle,
      hoverStartTime,
      hoverStayTime,
    } = this.props

    const isDefaultSize: boolean = size === 'default'
    const isDefaultType: boolean = type === 'default'
    const themeColorMap: { default: string[], primary: string[], warn: string[] } = {
      default: ['#F8F8F8', '#f7f7f7'],
      primary: ['#1AAD19', '#9ED99D'],
      warn: ['#E64340', '#EC8B89'],
    }
    // Use themeColorMap normally as PLAIN is false (by default),
    // otherwise use rgb(53,53,53) for plain-default-type particularly.
    const themeColor: string = plain && isDefaultType ? `rgba(53,53,53,${disabled ? 0.6 : 1})` : themeColorMap[type][disabled ? 1 : 0]
    const backgroundColor: string = plain ? 'transparent' : themeColor
    const borderStyle: StyleProp<ViewStyle> = plain && { borderWidth: 1, borderColor: themeColor }
    const textColor: string = plain
      ? themeColor
      : (isDefaultType ? `rgba(0,0,0,${disabled ? 0.3 : 1})` : `rgba(255,255,255,${disabled ? 0.6 : 1})`)

    const rotateDeg: Animated.AnimatedInterpolation = this.state.valve.interpolate({
      inputRange: [0, 1],
      outputRange: ['0deg', '360deg']
    })

    return (
      <TouchableWithoutFeedback
        delayPressIn={hoverStartTime}
        delayPressOut={hoverStayTime}
        onPress={this.onPress}
        onPressIn={this.onPressIn}
        onPressOut={this.onPressOut}
        ref={this.$touchable}
      >
        <View
          style={[
            styles.btn,
            !isDefaultSize && styles.btnMini,
            { backgroundColor },
            borderStyle,
            style,
            this.state.isHover && hoverStyle
          ]}
        >
          {loading && (
            <Animated.View
              style={[styles.loading, { transform: [{ rotate: rotateDeg }] }]}
            >
              <Image
                source={type === 'warn' ? require('../../assets/loading-warn.png') : require('../../assets/loading.png')}
                style={styles.loadingImg}
              />
            </Animated.View>)
          }
          {typeof children === 'string' ? <Text
            style={[
              styles.btnText,
              !isDefaultSize && styles.btnTextMini,
              { color: textColor }
            ]}
          >
            {children}
          </Text> : children}
        </View>
      </TouchableWithoutFeedback>
    )
  }
}

export default _Button
