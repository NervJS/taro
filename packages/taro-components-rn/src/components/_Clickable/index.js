/**
 * ✔ hoverStyle: Convert hoverClass to hoverStyle.
 * ✔ hoverStartTime
 * ✔ hoverStayTime
 * ✘ hoverStopPropagation: Fixed value TRUE
 * ✔ onClick(tap)
 * ✔ onLongPress(longpress)
 * ✔ onTouchstart
 * ✔ onTouchmove
 * ✔ onTouchcancel
 * ✔ onTouchend
 *
 * @flow
 */

import * as React from 'react'
import {
  View,
  // TouchableWithoutFeedback,
  StyleSheet,
  PanResponder,
} from 'react-native'
import { omit, dismemberStyle } from '../../utils'

type Props = {
  style?: StyleSheet.Styles,
  hoverStyle?: StyleSheet.Styles,
  hoverStartTime: number,
  hoverStayTime: number,
  // hoverStopPropagation?: boolean,
  onClick?: Function,
  onLongPress?: Function,
  onTouchstart?: Function,
  onTouchmove?: Function,
  onTouchcancel?: Function,
  onTouchend?: Function,
}

type State = {
  isHover: boolean
}

export default function (WrappedComponent: React.ComponentType<*>) {
  return class _Clickable extends React.Component<Props, State> {
    // eslint-disable-next-line no-useless-constructor
    constructor (props: Props) {
      super(props)
    }

    props: Props
    state: State = {
      isHover: false
    }
    startTimestamp: number = 0
    startTimer: TimeoutID
    stayTimer: TimeoutID

    panResponder: any = PanResponder.create({
      onStartShouldSetPanResponder: (evt, gestureState) => {
        const {
          hoverStyle,
          onClick,
          onLongPress,
          onTouchstart,
          onTouchmove,
          onTouchcancel,
          onTouchend
        } = this.props
        return hoverStyle || onClick || onLongPress || onTouchstart || onTouchmove || onTouchcancel || onTouchend
      },
      onMoveShouldSetPanResponder: (evt, gestureState) => {
        const { onTouchmove, onTouchcancel, onTouchend } = this.props
        return onTouchmove || onTouchcancel || onTouchend
      },
      onPanResponderGrant: (evt, gestureState) => {
        const { onTouchstart } = this.props
        onTouchstart && onTouchstart(this.getWxAppEvent(evt))
        this.startTimestamp = evt.nativeEvent.timestamp
        this.setStartTimer()
      },
      onPanResponderMove: (evt, gestureState) => {
        const { onTouchmove } = this.props
        onTouchmove && onTouchmove(this.getWxAppEvent(evt))
      },
      onPanResponderTerminationRequest: (evt, gestureState) => true,
      onPanResponderRelease: (evt, gestureState) => {
        const { onClick, onLongPress, onTouchend } = this.props
        onTouchend && onTouchend(this.getWxAppEvent(evt))
        const endTimestamp = evt.nativeEvent.timestamp
        const gapTime = endTimestamp - this.startTimestamp
        const hasMove = Math.abs(gestureState.dx) >= 1 || Math.abs(gestureState.dy) >= 1
        if (!hasMove) {
          if (gapTime <= 350) {
            onClick && onClick(this.getWxAppEvent(evt))
          } else {
            onLongPress && onLongPress(this.getWxAppEvent(evt))
          }
        }
        this.setStayTimer()
      },
      onPanResponderTerminate: (evt, gestureState) => {
        const { onTouchcancel } = this.props
        onTouchcancel && onTouchcancel(this.getWxAppEvent(evt))
        this.setStayTimer()
      },
    })

    static defaultProps = {
      hoverStartTime: 20,
      hoverStayTime: 70
    }

    setStartTimer = () => {
      const { hoverStyle, hoverStartTime } = this.props
      if (hoverStyle) {
        this.startTimer && clearTimeout(this.startTimer)
        this.startTimer = setTimeout(() => {
          this.setState({ isHover: true })
        }, hoverStartTime)
      }
    }

    setStayTimer = () => {
      const { hoverStyle, hoverStayTime } = this.props
      this.startTimer && clearTimeout(this.startTimer)
      if (hoverStyle) {
        this.stayTimer && clearTimeout(this.stayTimer)
        this.stayTimer = setTimeout(() => {
          this.startTimer && clearTimeout(this.startTimer)
          this.state.isHover && this.setState({ isHover: false })
        }, hoverStayTime)
      }
    }

    getWxAppEvent = (event: Object) => {
      const nativeEvent = event.nativeEvent
      const { timestamp, target, pageX, pageY, touches = [], changedTouches = [] } = nativeEvent
      return {
        type: 'tap',
        timeStamp: timestamp,
        target: {
          id: target,
          dataset: {}
        },
        currentTarget: {
          id: target,
          dataset: {}
        },
        detail: {
          x: pageX,
          y: pageY
        },
        touches: touches.map((item) => {
          return {
            identifier: item.identifier,
            pageX: item.pageX,
            pageY: item.pageY,
            clientX: item.locationX,
            clientY: item.locationY
          }
        }),
        changedTouches: changedTouches.map((item) => {
          return {
            identifier: item.identifier,
            pageX: item.pageX,
            pageY: item.pageY,
            clientX: item.locationX,
            clientY: item.locationY
          }
        })
      }
    }

    componentWillUnmount () {
      this.startTimer && clearTimeout(this.startTimer)
      this.stayTimer && clearTimeout(this.stayTimer)
    }

    render () {
      const {
        style,
        hoverStyle,
        // hoverStopPropagation,
        onClick,
        onLongPress,
        onTouchstart,
        onTouchmove,
        onTouchcancel,
        onTouchend,
      } = this.props
      const { isHover } = this.state

      if (
        !hoverStyle &&
        // !hoverStopPropagation &&
        !onClick &&
        !onLongPress &&
        !onTouchstart &&
        !onTouchmove &&
        !onTouchcancel &&
        !onTouchend
      ) {
        return (
          <WrappedComponent {...this.props} />
        )
      }

      const dismember = dismemberStyle(style)
      const hoverStyleDismember = isHover ? dismemberStyle(hoverStyle) : {}

      return (
        <View
          {...this.panResponder.panHandlers}
          style={[dismember.wrapperStyle, hoverStyleDismember.wrapperStyle]}
        >
          <WrappedComponent
            {...omit(this.props, [
              'style',
              'hoverStyle',
              'hoverStartTime',
              'hoverStayTime',
              'onClick',
              'onLongPress',
              'onTouchstart',
              'onTouchmove',
              'onTouchcancel',
              'onTouchend'
            ])}
            style={[dismember.innerStyle, hoverStyleDismember.innerStyle]}
          />
        </View>
      )
    }
  }
}
