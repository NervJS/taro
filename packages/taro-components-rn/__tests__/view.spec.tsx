import * as React from 'react'
// import 'react-native'
import { _View } from '../src/components/view'
import * as renderer from 'react-test-renderer'

describe('<View />', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <_View />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
