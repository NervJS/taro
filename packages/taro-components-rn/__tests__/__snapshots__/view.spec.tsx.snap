// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<View /> renders correctly 1`] = `<View />`;
